### PR TITLE
Port do_afters

### PR DESCRIPTION
**Pull request by aidan-bowman**
## About the Pull Request

Ports do_after improvements from TG.
Ports actionspeed modifiers from TG (will be implemented in #1226).
Ports TG sprites.
Ports interactive do_afters.
Also I went through all do_afters and fixed em up.

Credit to TG and Mojave Sun (technically).

## Why It's Good For The Game

Alright alright I get it, I can hear it already, but trust me there's a few good reasons to have interactive do_afters

1. Requires people to actually focus on things that would normally require focus. IRL it's pretty hard to operate a control panel while also staring at a massive statue, so why is it so easy in-game?
2. Gives us more variables to tweak with do_afters. It's not just "you go faster or slower", you can manipulate reward values, QTE difficulty, speed, hell you could even have wacky things (being high, going insane, memetic agents) have wacky effects (moving targets, fake targets, alternate coloring, etc).
3. Lets us add in more interesting gameplay for do_afters. Loads of servers have do_afters where you want to do something that someone else wants you to NOT do (forcefeeding, breaking handcuffs, etc), and usually the only real "game" there is to swap places with them and they magically stop (both unrealistic and quite boring). Instead, we can try and have mechanics where you actually have to STOP the person trying to inject you instead of just abusing the system in the dumbest way possible.

## Changelog

:cl:
add: Interactive do_afters! During (most) timed actions, stars will pop up that can be clicked to get extra progress.
add: Actionspeed modifiers ported from TG.
Balance: do_afters using the interactive system have had default times increased to account for extra progress.
imageadd: do_after sprites ported from TG.
imageadd: Interactive do_after sprites by MrDoomBringer.
refactor: Massive improvements to do_after system (from TG).
/:cl:

### CODE REVIEW

**Comment by aidan-bowman** [comment]
Had to delete and re-open PR to resolve merge conflict.

**Comment by Tsurupeta** [comment]
Bro can't resolve merge conflicts without closing his PRs...

**Comment by TheDarkElites** [comment]
[13:33:52] Runtime in progressbar.dm, line 103: bad index
proc name: Destroy (/datum/progressbar/Destroy)
src: /datum/progressbar (/datum/progressbar)
call stack:
/datum/progressbar (/datum/progressbar): Destroy(0)
qdel(/datum/progressbar (/datum/progressbar), 0)
/datum/progressbar (/datum/progressbar): qdel self()
/datum/callback (/datum/callback): InvokeAsync()
Timer (/datum/controller/subsystem/timer): fire(0)
Timer (/datum/controller/subsystem/timer): ignite(0)
Master (/datum/controller/master): RunQueue()
Master (/datum/controller/master): Loop()
Master (/datum/controller/master): StartProcessing(0)

[14:37:49] Runtime in do_afters.dm, line 23: do_after was passed a non-number delay: null.
proc name: do after (/proc/do_after)
usr: SCP-049 (/mob/living/carbon/human/scp049) ([0x3000018]) (the floor) (195,194,1) (/turf/simulated/floor/tiled/steel_grid) (k4rl)
usr.loc: The floor (/turf/simulated/floor/tiled/steel_grid) ([0x100c101]) (195,194,1) (/area/site53/lhcz/scp049containment)
src: null
call stack:
do after(10, null, null, 330, 61, null, null, 1, 100, 'sounds/machines/click.ogg', 10)
SCP-049 (/mob/living/carbon/human/scp049): Pry Open Airlock(049 Access (/obj/machinery/door/airlock/glass/security))

[14:36:21] Runtime in progressbar_booster.dm, line 46: Cannot modify null.alpha.
proc name: Click (/obj/screen/progressbar_booster/Click)
usr: Unknown (as Ingwe Styles) (/mob/living/carbon/human) ([0x3000089]) (the floor) (112,115,4) (/turf/simulated/floor/tiled/monotile/white) (ingwe)
usr.loc: The floor (/turf/simulated/floor/tiled/monotile/white) ([0x1036c00]) (112,115,4) (/area/site53/upper_surface/commstower)
src: the focus (/obj/screen/progressbar_booster)
src.loc: null
call stack:
the focus (/obj/screen/progressbar_booster): Click(the floor (112,115,4) (/turf/simulated/floor/tiled/monotile/white), "mapwindow.map", "icon-x=17;icon-y=19;left=1;but...")
Ingwe (/client): Click(the focus (/obj/screen/progressbar_booster), the floor (112,115,4) (/turf/simulated/floor/tiled/monotile/white), "mapwindow.map", "icon-x=17;icon-y=19;left=1;but...")

**Comment by aidan-bowman** [comment]
Major code rewrite necessary :/

**Comment by TheDarkElites** [comment]
truly a guh moment

**Comment by aidan-bowman** [comment]
Code rewrite done

**Review by TheDarkElites — COMMENTED** [review]
You should probably go through and implement the new image holder datum wherever else it may be necessary or at the very least make a separate pr later to do this.

**Comment by aidan-bowman** [comment]
Alright, I'm 93.7% sure this is working now

**Comment by TheDarkElites** [comment]
[15:28:08] Runtime in progressbar.dm, line 103: bad index
proc name: Destroy (/datum/progressbar/Destroy)
src: /datum/progressbar (/datum/progressbar)
call stack:
/datum/progressbar (/datum/progressbar): Destroy(0)
qdel(/datum/progressbar (/datum/progressbar), 0)
/datum/progressbar (/datum/progressbar): qdel self()
/datum/callback (/datum/callback): InvokeAsync()
Timer (/datum/controller/subsystem/timer): fire(0)
Timer (/datum/controller/subsystem/timer): ignite(0)
Master (/datum/controller/master): RunQueue()
Master (/datum/controller/master): Loop()
Master (/datum/controller/master): StartProcessing(0)

File: code/datums/progressbar_booster.dm [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+	booster.image_pixel_x = x
+	booster.image_pixel_y = y
+
+	trap.image_pixel_x = x
+	trap.image_pixel_y = y
+
+	booster.regenerate_image()
+	trap.regenerate_image()
+
+	spawn()
```
Comment (Tsurupeta) [inline code/datums/progressbar_booster.dm]: why spawn() if you can use INVOKE_ASYNC

Comment (Tsurupeta) [reply code/datums/progressbar_booster.dm]: or even timer?

File: code/modules/actionspeed/_actionspeed_modifier.dm [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+
+	/// Other modification datums this conflicts with.
+	var/conflicts_with
+
+/datum/actionspeed_modifier/New()
+	. = ..()
+	if(!id)
+		id = "[type]" //We turn the path into a string.
+
+GLOBAL_LIST_EMPTY(actionspeed_modification_cache)
```
Comment (Tsurupeta) [inline code/modules/actionspeed/_actionspeed_modifier.dm]: move this to the top of file idk, don't declare global lists in middle of file

Comment (aidan-bowman) [reply code/modules/actionspeed/_actionspeed_modifier.dm]: done

File: code/datums/progressbar_booster.dm (hunk 1 above)
Comment (aidan-bowman) [reply code/datums/progressbar_booster.dm]: done

**Review by Tsurupeta — APPROVED** [review]

**Pull request merged**